File: arrow-ipc/ipc-rs/src/write_ipc.rs
//! file: write_ipc.rs
//! author: Jacob Xie
//! date: 2023/04/28 16:51:19 Friday
//! brief:

use std::fs::File;
use std::net::TcpStream;

use arrow2::array::Array;
use arrow2::chunk::Chunk;
use arrow2::datatypes::Schema;
use arrow2::error::Result;
use arrow2::io::ipc::write;

pub fn write_batches(path: &str, schema: Schema, chunks: &[Chunk<Box<dyn Array>>]) -> Result<()> {
    let file = File::create(path)?;

    let options = write::WriteOptions { compression: None };
    let mut writer = write::FileWriter::new(file, schema, None, options);

    writer.start()?;
    for chunk in chunks {
        writer.write(chunk, None)?;
    }
    writer.finish()
}

// Streaming write
pub fn write_stream(addr: &str, schema: Schema, chunks: &[Chunk<Box<dyn Array>>]) -> Result<()> {
    let mut writer = TcpStream::connect(addr)?;
    let mut stream =
        write::StreamWriter::new(&mut writer, write::WriteOptions { compression: None });

    stream.start(&schema, None)?;
    for chk in chunks {
        stream.write(chk, None)?;
    }
    stream.finish()?;

    Ok(())
}

File: arrow-ipc/ipc-cpp/write_ipc.cpp
/**
 * @file:	write_ipc.cpp
 * @author:	Jacob Xie
 * @date:	2023/04/28 15:17:09 Friday
 * @brief:
 **/

#include "write_ipc.h"

// ================================================================================================
//
// ================================================================================================

arrow::Status write_ipc_file(std::string filename, std::shared_ptr<arrow::Table> table)
{
  ARROW_ASSIGN_OR_RAISE(auto outfile, arrow::io::FileOutputStream::Open(filename));
  ARROW_ASSIGN_OR_RAISE(auto writer, arrow::ipc::MakeFileWriter(outfile, table->schema()));
  ARROW_RETURN_NOT_OK(writer->WriteTable(*table));
  ARROW_RETURN_NOT_OK(writer->Close());

  return arrow::Status::OK();
}
